boards: add Adafruit Feather nrf52840 Express UF2

jacobw opened the pull request:
Add support for flashing the Adafruit Feather nrf52840 Express using UF2 over USB.

This PR includes converting the boards to variants to be consistent with hwmv2 and remove unnecessary unique boards.

jacobw committed:
boards: adafruit feather nrf52840 express: add UF2

Add support for flashing the Adafruit Feather nrf52840 Express
using UF2 over USB.

Adding a new board appears to be the best way to add this as it
requires changes to Kconfig, which can only match on board name.

Signed-off-by: Jacob Winther <jacob@9.nz>

jadejr commented:
The xiao_ble board  defaults to uf2.  They chose uf2 by default since that's how the hardware ships. I'm guessing the feather defaults to utf2, but default is reversed. It'd be nice if there was a common approach here across all zephyr boards.

jacobw commented:
> The xiao_ble board defaults to uf2.

the xiao_ble doesn’t have an SWD connector does it? I think the current approach is default to SWD if it has a connector, and UF2 if not (and pre flashed with the boot loader).

the Sense has SWD headers but no connector, so only has UF2 support.

jadejr commented:
> > The xiao_ble board defaults to uf2.
> 
> the xiao_ble doesn’t have an SWD connector does it? I think the current approach is default to SWD if it has a connector, and UF2 if not (and pre flashed with the boot loader).
> 
Both versions have swd pins on the bottom which you can use with their expansion board or pogo pins.  I'm using their expansion board plus a nordic dev kit to flash the non-Sense version. You can see them right here 
![image](https://github.com/user-attachments/assets/93442370-a500-4258-97ff-c76af551c815)

jacobw commented:
> > > The xiao_ble board defaults to uf2.
> > 
> > 
> > the xiao_ble doesn’t have an SWD connector does it? I think the current approach is default to SWD if it has a connector, and UF2 if not (and pre flashed with the boot loader).
> 
> Both versions have swd pins

That's why I suggested connector. Sense has pads but not headers, which would be even harder to use. 

Overall, I couldn't find an easy way to support both without having multiple boards. I was hoping to do something cleaner with the variants in hwmv2, but that didn't seem possible. Not sure there is a good answer to this.

jadejr commented:
I'm just suggesting whatever the answer is should be applied to all boards that default to utf2.

jacobw commented:
> I'm just suggesting whatever the answer is (good or not) should be applied to all boards that default to utf2.

I've been wondering what the options are for dealing with this in a consistent way.

I thought  UF2 required specific board level definitions to set the Kconfig options, but looking xiao_ble_defconfig perhaps that can all be managed as board variants. In that case, perhaps an approach is to always default to SWD, and have a UF2 variant for any board that needs/wants it.

That would mean creating new non-UF2 definitions and that would require maintenance so perhaps that is too much effort in the long term. Or does that just mean people will be creating custom boards out of tree?

jacobw committed:
boards: adaftuit feather nrf52840 Kconfig updates

Remove unnecessary Kconfig options.
Simplify configuration by removing custom Kconfig option.
Make Kconfig generic for all boards by moving usb selection to
board_defconfig.

These changes should enable the boards to become variants instead.

Signed-off-by: Jacob Winther <jacob@9.nz>

jacobw committed:
boards: adafruit feather nrf52840 remove bossac

Remove bossac support for UF2 as it isn't neccessary.

Based on xiao_ble.

Signed-off-by: Jacob Winther <jacob@9.nz>

jacobw committed:
boards: adafruit feather nrf52840 convert to variants

Update all versions of adafruit feather nrf52840 to be variants rather
than separate boards. The aim is to provide a more consistent approach
and to remove some unnecessary Kconfig redundancy.

This commit also tries to stanardise on using a uf2 variant for uf2
support as most boards with uf2 have swd headers so its helpful to
have support for both options.

This commit seperates the two common dts setups into dtsi to avoid
duplicating the somewhat cryptic flash layouts. I'm not sure if there
is a better way to do this.

Signed-off-by: Jacob Winther <jacob@9.nz>

jacobw committed:
boards: adafruit feather nrf52840 update docs

Update documentation to reflect move to variants

Signed-off-by: Jacob Winther <jacob@9.nz>

jacobw committed:
boards: adafruit feather nrf52850 update adc test overlays

Update overlays for adc tests.

Signed-off-by: Jacob Winther <jacob@9.nz>

jacobw commented:
PR updated as per the approach discussed above. Boards converted to variants.

jacobw committed:
boards: adafruit feather nrf52840 update depricated

Update depricated.cmake to reflect new name for this board.

Signed-off-by: Jacob Winther <jacob@9.nz>

jacobw commented:
Ping @anangl @decsny  as per PR Review Escalation. It's been a week since all checks passed.

decsny commented:
im not actually sure why this got assigned to the ADC area

decsny commented:
added @nordicjm as he has been reviewing hwmv2 board rework type stuff

nordicjm commented on boards/adafruit/feather_nrf52840/Kconfig:
```diff
@@ -1,12 +1,12 @@
 
-if BOARD_ADAFRUIT_FEATHER_NRF52840_SENSE
+if (BOARD_ADAFRUIT_FEATHER_NRF52840_SENSE || BOARD_ADAFRUIT_FEATHER_NRF52840_EXPRESS_UF2)
 config BOARD_SERIAL_BACKEND_CDC_ACM
```
```suggestion
if(BOARD_ADAFRUIT_FEATHER_NRF52840_SENSE || BOARD_ADAFRUIT_FEATHER_NRF52840_EXPRESS_UF2)

config BOARD_SERIAL_BACKEND_CDC_ACM
```

nordicjm commented on boards/adafruit/feather_nrf52840/adafruit_feather_nrf52840_express_uf2.dts:
```diff
@@ -0,0 +1,72 @@
+		reserved_partition_0: partition@0 {
+			label = "SoftDevice";
+			reg = <0x000000000 DT_SIZE_K(152)>;
+		};
```
would add `read-only;`

nordicjm commented on boards/adafruit/feather_nrf52840/adafruit_feather_nrf52840_express_uf2.dts:
```diff
@@ -0,0 +1,72 @@
+		boot_partition: partition@f4000 {
+			label = "UF2";
+			reg = <0x000f4000 DT_SIZE_K(48)>;
+		};
```
likewise here too

nordicjm commented on boards/adafruit/feather_nrf52840/adafruit_feather_nrf52840_express_uf2_defconfig:
```diff
@@ -0,0 +1,21 @@
+# Enable MPU
+CONFIG_ARM_MPU=y
+
+# enable GPIO
```
```suggestion
# Enable GPIO
```

nordicjm commented on boards/adafruit/feather_nrf52840/adafruit_feather_nrf52840_express_uf2_defconfig:
```diff
@@ -0,0 +1,21 @@
+# enable GPIO
+CONFIG_GPIO=y
+
+# enable uart driver
```
```suggestion
# Enable UART driver
```

nordicjm commented on boards/adafruit/feather_nrf52840/adafruit_feather_nrf52840_express_uf2_defconfig:
```diff
@@ -0,0 +1,21 @@
+# enable uart driver
+CONFIG_SERIAL=y
+
+# enable console
```
```suggestion
# Enable console
```

nordicjm commented on boards/adafruit/feather_nrf52840/adafruit_feather_nrf52840_express_uf2_defconfig:
```diff
@@ -0,0 +1,21 @@
+CONFIG_CONSOLE=y
+CONFIG_UART_CONSOLE=y
+
+# enable bossac
```
```suggestion
# Enable bossac
```

nordicjm commented on boards/adafruit/feather_nrf52840/board.cmake:
```diff
@@ -2,6 +2,9 @@
 board_runner_args(jlink "--device=nRF52840_xxAA" "--speed=4000")
 board_runner_args(pyocd "--target=nrf52840" "--frequency=4000000")
+if(CONFIG_BOARD_ADAFRUIT_FEATHER_NRF52840_EXPRESS_UF2)
+include(${ZEPHYR_BASE}/boards/common/uf2.board.cmake)
```
2 space indent for cmake

nordicjm commented on boards/adafruit/feather_nrf52840/board.cmake:
```diff
@@ -2,6 +2,9 @@
+if(CONFIG_BOARD_ADAFRUIT_FEATHER_NRF52840_EXPRESS_UF2)
+include(${ZEPHYR_BASE}/boards/common/uf2.board.cmake)
+endif()
 if(CONFIG_BOARD_ADAFRUIT_FEATHER_NRF52840_SENSE)
```
elseif()

nordicjm commented on boards/adafruit/feather_nrf52840/doc/index.rst:
```diff
@@ -142,6 +142,10 @@ The Feather nRF52840 ships with the `Adafruit nRF52 Bootloader`_ which
 
+To flash the ``adafruit_feather_nrf52480_express`` using UF2, you must use
+the ``adafruit_feather_nrf52480_express_uf2`` board, which uses an appropriate
+flash layout.
```
I don't get it, if this is actually the same board, then this should be a variant, not a new board

nordicjm commented:
> Add support for flashing the Adafruit Feather nrf52840 Express using UF2 over USB.

Having see this just now, yes this is a wrong design, it needs to be a variant of the existing board

jacobw commented:
> > Add support for flashing the Adafruit Feather nrf52840 Express using UF2 over USB.
> 
> Having see this just now, yes this is a wrong design, it needs to be a variant of the existing board

Thanks for the review. It took my a long time to realise that BOARD_$(BOARD_TARGET_STRING) was an option, but I work around that and converted everything to variants. Is that what you mean. I see I missed that in the documentation so will fix,

nordicjm commented:
> > > Add support for flashing the Adafruit Feather nrf52840 Express using UF2 over USB.
> > 
> > 
> > Having see this just now, yes this is a wrong design, it needs to be a variant of the existing board
> 
> Thanks for the review. It took my a long time to realise that BOARD_$(BOARD_TARGET_STRING) was an option, but I work around that and converted everything to variants. Is that what you mean. I see I missed that in the documentation so will fix,

Best way to see it would be an existing board, so https://github.com/zephyrproject-rtos/zephyr/tree/main/boards/nordic/nrf54l15pdk is an example, if you check the board.yml file you will see:
```
board:
  name: nrf54l15pdk
  vendor: nordic
  socs:
  - name: nrf54l15
    variants:
    - name: xip
      cpucluster: cpuflpr
```

This means that there is a board nrf54l15pdk, it has a soc nrf54l15, this soc has multiple CPU cores, for the cpuflpr core, you can optionally build just for that or you can build for the XIP variant of that, if you build for the XIP variant, it uses the other configuration

jacobw commented:
> Best way to see it would be an existing board

I convert to variants in a later commit in this PR. Sorry if that was confusing, I covert the existing boards to variants in the same commit.

jacobw commented:
@nordicjm I realise I've structured this PR poorly. I'll restructure it and submit it as a new PR. Apologies for wasting your time.
